go/libraries/events/file_backed_proc.go

```diff
@@ -0,0 +1,186 @@
+func MD5FileNamer(bytes []byte) string {
+	return MD5StrUrl(bytes) + evtDataExt
+}
+
+type eventsDataDir struct {
+	fs   filesys.Filesys
+	path string
+}
+
+func newEventsDataDir(fs filesys.Filesys, homeDir string, doltDir string) *eventsDataDir {
+	path := filepath.Join(homeDir, doltDir, eventsDir)
+
+	return &eventsDataDir{fs: fs, path: path}
+}
+
+//  MakeEventsDir creates a new events data dir in the main dolt dir
+func (evd *eventsDataDir) MakeEventsDir() error {
+	if exists, _ := evd.fs.Exists(evd.path); !exists {
+		if err := evd.fs.MkDirs(evd.path); err != nil {
+			return err
+		}
+		return nil
+	}
+	return nil
+}
+
+func (evd *eventsDataDir) getPath() string {
+	return evd.path
+}
+
+// FileBackedProc writes events requests to files in an events data dir
+type FileBackedProc struct {
+	ed         *eventsDataDir
+	namingFunc fileNamingFunc
+}
+
+// NewFileBackedProc creates a new FileBackedProc
+func NewFileBackedProc(fs filesys.Filesys, userHomeDir string, doltDir string, nf fileNamingFunc) *FileBackedProc {
+	eventsDataDir := newEventsDataDir(fs, userHomeDir, doltDir)
+
+	return &FileBackedProc{ed: eventsDataDir, namingFunc: nf}
+}
+
+func (fbp *FileBackedProc) renameFile(dir string, oldName string) error {
+	oldPath := filepath.Join(dir, oldName)
+
+	data, err := fbp.ed.fs.ReadFile(oldPath)
+	if err != nil {
+		return err
+	}
+
+	filename := fbp.namingFunc(data)
+	newPath := filepath.Join(dir, filename)
+
+	if err := fbp.ed.fs.MoveFile(oldPath, newPath); err != nil {
+		return nil
+	}
+
+	return nil
+}
+
+// EventsDirExists returns true iff the events data dir exists
+func (fbp *FileBackedProc) EventsDirExists() bool {
+	if exists, _ := fbp.ed.fs.Exists(fbp.ed.getPath()); exists {
+		return true
+	}
+	return false
+}
+
+// GetEventsDirPath returns the path to the events data dir
+func (fbp *FileBackedProc) GetEventsDirPath() string {
+	return fbp.ed.getPath()
+}
+
+// GetFileSys returns the current filesys being used
+func (fbp *FileBackedProc) GetFileSys() filesys.Filesys {
+	return fbp.ed.fs
+}
+
+// WriteEvents writes events requests to the events data dir
+func (fbp *FileBackedProc) WriteEvents(version string, evts []*eventsapi.ClientEvent) error {
+	if len(evts) < 1 {
+		return nil
+	}
+
+	var plat eventsapi.Platform
+	switch strings.ToLower(runtime.GOOS) {
+	case "darwin":
+		plat = eventsapi.Platform_DARWIN
+	case "linux":
+		plat = eventsapi.Platform_LINUX
+	case "windows":
+		plat = eventsapi.Platform_WINDOWS
+	}
+
+	if err := fbp.ed.MakeEventsDir(); err != nil {
+		return err
+	}
+
+	eventsPath := fbp.ed.getPath()
+	tempFilename := filepath.Join(eventsPath, localPath)
+
+	f, err := fbp.ed.fs.OpenForWrite(tempFilename)
+
+	if err != nil {
+		return err
+	}
+
+	req := &eventsapi.LogEventsRequest{
+		MachineId: getMachineID(),
+		Version:   version,
+		Platform:  plat,
+		Events:    evts,
+	}
+
+	data, err := proto.Marshal(req)
+	if err != nil {
+		return err
+	}
+
+	if _, err := f.Write(data); err != nil {
```

Comment:
it's possible that Write will write less than the entire buffer.  iohelp.WriteAll will take care of flushing the entire contents of data.